### ISSUE TITLE
batch size changes after update_metric

### ISSUE THREAD

**Issue by Rui-Zhou-2** [lorenmt/auto-lambda#5]
Hi, I have a small question here, why the batch size of the model output changes after this line:

https://github.com/lorenmt/auto-lambda/blob/24591b7ff0d4498b18bd4b4c85c41864bdfc800a/trainer_dense.py#L214

it seems before this line the batch is 4, then it comes to 3.

since I want to use the output of the model after` loss.backward()` in the next epoch...so it becomes a problem if the batch size change.

would you kindly give me some idea on this?

With best regards

**Comment by lorenmt** [comment]
Hello, could you be more specifIed which tensor changed the batch size?
This is super strange as if you check out the update_metric function, it does not modify train_pred or train_target in any way.

One only possible related issue I can imagine is the batch size would be different in the last batch of data due to the divisibility problem. But it would be changed right after we fetch the data not in the update_metric function.

**Comment by Rui-Zhou-2** [comment]
Batch of data may be the right cause,  but I can not find where it will change back.

Like I select the `train_pred[0]`, and the shape changed from [4,13,H,W] to [3,13,H,W] 

from    `test_metric.update_metric(test_pred, test_target, test_loss)`,

until the end https://github.com/lorenmt/auto-lambda/blob/24591b7ff0d4498b18bd4b4c85c41864bdfc800a/trainer_dense.py#L237

**Comment by lorenmt** [comment]
I will verify that and get back to you later.

**Comment by lorenmt** [comment]
Hello, I added the following code:
```
print(f'train batch before metric: {train_pred[0].shape}')
train_metric.update_metric(train_pred, train_target, train_loss)
print(f'train batch after metric: {train_pred[0].shape}')

print(f'test batch before metric: {test_pred[0].shape}')
test_metric.update_metric(test_pred, test_target, test_loss)
print(f'test batch after metric: {test_pred[0].shape}')
```

And I was not able to reproduce your problem. Both training and testing have the same batch size 4 except for the last batch of data which is 3 (due to the divisibility problem, which can be turned off by adding drop_last=True in data loader), but they are consistent before and after update metric function.

**Comment by Rui-Zhou-2** [comment]
With great thanks👍
drop_last=True solved this issue.
